src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +550,123 @@ ApplicationImpl::joinAllThreads()
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Invalid manual close ledger "
+                                       "sequence number {} (minimum {})"),
+                            targetLedgerSeq, minLedgerSeq));
+        }
+        auto const lastCloseTime = getLedgerManager()
+                                       .getLastClosedLedgerHeader()
+                                       .header.scpValue.closeTime;
+        auto const now = VirtualClock::to_time_t(getClock().system_now());
+        uint32_t const defaultManualCloseTimeInterval = 5;
+        auto nextCloseTime = lastCloseTime + 1;
+        auto const defaultNextCloseTime =
+            lastCloseTime + defaultManualCloseTimeInterval;
+        if (nextCloseTime < now)
+        {
+            nextCloseTime = now;
+        }
+        if (manualCloseTime)
+        {
+            if (*manualCloseTime < nextCloseTime)
+            {
+                throw std::invalid_argument(
+                    fmt::format(FMT_STRING("Invalid manual close time {} (last "
+                                           "close time={}, now={})"),
+                                *manualCloseTime, lastCloseTime, now));
+            }
+            nextCloseTime = *manualCloseTime;
+        }
+        else
+        {
+            if (nextCloseTime < defaultNextCloseTime)
+            {
+                nextCloseTime = defaultNextCloseTime;
+            }
+        }
+        if (targetLedgerSeq != minLedgerSeq)
+        {
+            if (!mHerder->isTransactionQueueEmpty())
+            {
+                throw std::invalid_argument(
+                    fmt::format(FMT_STRING("May not skip ledgers (requested "
+                                           "sequence number {}; next {}) while "
+                                           "transaction queue is non-empty"),
+                                targetLedgerSeq, minLedgerSeq));
+            }
+            LOG(INFO) << "closing ledger with sequence number "
+                      << targetLedgerSeq - 1 << " on behalf of target "
+                      << targetLedgerSeq;
+            LedgerTxn ltx(getLedgerTxnRoot());
+            ltx.loadHeader().current().ledgerSeq = targetLedgerSeq - 1;
+            getLedgerManager().manuallyAdvanceLedgerHeader(
+                ltx.loadHeader().current());
+            ltx.commit();
+            auto const& header =
+                getLedgerManager().getLastClosedLedgerHeader().header;
+            dynamic_cast<HerderImpl&>(getHerder())
+                .getHerderSCPDriver()
+                .restoreSCPState(header.ledgerSeq, header.scpValue);
+        }
+        if (now < nextCloseTime)
+        {
+            getClock().setCurrentVirtualTime(
+                VirtualClock::time_point(std::chrono::seconds(nextCloseTime)));
+        }
+        mHerder->triggerNextLedger(targetLedgerSeq, true);
+        while (getClock().crank(true) != 0 &&
```

Comment:
This doesn't look right:
we should not call `crank` from here as this code is already being called from the main `crank` and I don't think we make any guarantees that this function is re-entrant.

Because `triggerNextLedger` posts an event, things should "just work", did you have any problems without this code?

Reply:
> did you have any problems without this code?

No, it _does_ just work.  This was me not adequately understanding how `VIRTUAL_TIME` worked.  I thought we'd have to manually `crank` to process the event(s) posted by `triggerNextLedger()` and the rest of the SCP processing resulting therefrom.  I've removed the `crank` loop.